### DISCUSSION TITLE
ClaRa: Is connect-equation: RealOuput connected with output variable of a connector legal?

### DISCUSSION THREAD

**Discussion by AlesVojacek** [OpenModelica/OpenModelica#8484] [General]
### Description

Is connect-equation: RealOuput connected with output variable of a connector legal? 
In Dymola, the example ClaRa.Visualisation.Check.TestQuadruple, where this is used, is working. But in but OpenModelica not. It says in OM eyeOut.m_flow is not a valid connector.

### Steps to Reproduce

See [example](https://libraries.openmodelica.org/branches/master/ClaRa_dev/files/ClaRa_dev_ClaRa.Visualisation.Check.TestQuadruple.err)
 
or have this minimalistic model

```mo
connector C
  output Real m_flow;
end C;

model M
  Modelica.Blocks.Sources.Ramp m;
equation
  connect(m.y,C.m_flow);
end M;`
```
### Version and OS

 - OpenModelica Version: OpenModelica v1.19.0-dev-541-gebaac9b11c (64-bit)
 - OS: [e.g. Windows 10, 64 bit]
 - Versions of used Modelica libraries if applicable

**Comment by perost** [comment]
`connect` takes two references to connectors, and `m_flow` is not a connector. I don't know why Dymola allows that, but I can't find anything in the specification that would allow using `connect` on something that's not a connector.

Note that `Modelica.Blocks.Interfaces.RealOutput` _is_ a connector, i.e. it's declared as:
```Modelica
connector RealOutput = output Real;
```
But in the model and your example you're not using `RealOutput`, just a normal non-connector `output Real`.

Edit: `m.y` is a `RealOutput` though, I guess that's what you meant.

**Comment by AlesVojacek** [comment]
You are right. Hence I just omit the connect-equation and just write an equation:
```
m.y=C.m_flow;
```
Then i works in OM as well.

**Comment by sjoelund** [comment]
Wouldn't it be better to declare the output Real as connector instead?

**Comment by AlesVojacek** [comment]
I have actually more variables then just m_flow (also p, h, T and s) so I do not want to have each of this variable to be a connector.

**Comment by casella** [comment]
I checked the original model `ClaRa.Visualisation.Check.TestQuadruple`. Indeed, the connect statements are invalid (the arguments need to be connectors), and substituting the connect statements such as `connect(m.y, eyeOut.m_flow)` with `m.y = eyeOut.m_flow` is indeed one way to fix the problem.

@AlesVojacek, what @sjoelund means is that if you want to be able to connect to individual variables in the eyeOut connector using the GUI, instead of manually writing equations (mixing graphical connections and textual equations can be confusing if you want to figure out how the model is built), you should declare a hierarchical connector instead, e.g.
```modelica
connector EyeOut "Signal bus featuring pressure, specific enthalpy,temperature, specific entropy and mass flow rate"
  Modelica.Blocks.Interfaces.RealOutput p "Pressure in bar";
  Modelica.Blocks.Interfaces.RealOutput h "Specific enthalpy in kJ/kg" annotation(HideResult=false);
  Modelica.Blocks.Interfaces.RealOutput m_flow "Mass flow rate in kg/s" annotation(HideResult=false);
  Modelica.Blocks.Interfaces.RealOutput T "Tempearture in degC" annotation (HideResult=false);
  Modelica.Blocks.Interfaces.RealOutput s "Specific entropy in kJ/kgK" annotation(HideResult=false);
end EyeOut;
``` 
Now p, h, m_flow, etc. are themselves connectors, so with this connector definition `connect(m.y, eyeOut.m_flow)` would become valid. When you connect the hierarchical connector, e.g. in `connect(quadruple.eye, eyeOut)`, the connect statement is automatically expanded into individual connect statements for each sub-connector.

**Comment by AlesVojacek** [comment]
I see your point now @sjoelund and @casella . I was not aware of the hierarchical connector. This allows the eyeOut connector, using the GUI, connect to individual variables, indeed. For clarity, should not then also the connector, which is used in quadruple, be rewritten as a hierarchical connector? That means:

```modelica
connector EyeIn "Signal bus featuring pressure, specific enthalpy, temperature, specific entropy and mass flow rate"
   Modelica.Blocks.Interfaces.RealInput p "Pressure in bar" annotation(HideResult=false);
   Modelica.Blocks.Interfaces.RealInput h "Specific enthalpy in kJ/kg" annotation(HideResult=false);
   Modelica.Blocks.Interfaces.RealInput m_flow "Mass flow rate in kg/s" annotation(HideResult=false);
   Modelica.Blocks.Interfaces.RealInput T "Tempearture in degC" annotation (HideResult=false);
   Modelica.Blocks.Interfaces.RealInput s "Specific entropy in kJ/kgK" annotation(HideResult=false);
end EyeIn;
```
instead of the existing:

```modelica
connector EyeIn "Signal bus featuring pressure, specific enthalpy, temperature, specific entropy and mass flow rate"
   import SI = ClaRa.Basics.Units;
   input Real p "Pressure in bar" annotation(HideResult=false);
   input Real h "Specific enthalpy in kJ/kg" annotation(HideResult=false);
   input Real m_flow "Mass flow rate in kg/s" annotation(HideResult=false);
   input Units.Temperature_DegC T "Tempearture in degC" annotation (HideResult=false);
   input Real s "Specific entropy in kJ/kgK" annotation(HideResult=false);
end EyeIn;
```
I mean, I tested both variants and did not see much of difference. They both works.

**Comment by sjoelund** [comment]
I think the existing is correct. It has units. I'm not sure if adding connector inside the connector matters. I would make them both declared in the analogous way though.

**Comment by AlesVojacek** [comment]
Now I am confused with what you @sjoelund  mean by existing and make them both in analogous way. Can you for clarity write how you think the EyeIn and Eyeout connectors could look like?

**Comment by sjoelund** [comment]
If EyeOut uses Modelica.Blocks use the same for EyeIn. If it uses output Real, use input Real for EyeIn. Then it looks consistent.

**Comment by AlesVojacek** [comment]
Right. The way how it was implemented was consistent, i.e. it used output Real for EyeOut connector and input Real for EyeIn connector. But then only way to make the ClaRa.Visualisation.Check.TestQuadruple works is to get rid off the connect equations: 
```modelica
connect(m.y, eyeOut.m_flow)
connect(T.y, eyeOut.T)
```
etc.
and write this instead:
```modelica
eyeOut.m_flow=m.y;
T.m_flow=T.y;
```
etc.
which is fine for me. But then you (@sjoelund) and @casella came up with the hierarchical connector, which would also work and would look more neat in this example ClaRaPlus.Visualisation.Check.TestQuadruple (since there would not mixing graphical connections and textual equations ). But I do not think there is such advantage of that. Hence, I will stick to output Real for EyeOut connector and input Real for EyeIn connecter as it is now and correct the ClaRaPlus.Visualisation.Check.TestQuadruple accordingly. Thanks a lot for the discussion @sjoelund and @casella .

**Comment by perost** [comment]
I'm closing this since it's not an issue with OpenModelica that needs to be fixed, but feel free to continue the discussion if you want to.

**Comment by casella** [comment]
@AlesVojacek, I'm not sure why @sjoelund brought up using input Real variables instead of RealInput connectors inside the connector, since he actually proposed the opposite in a [previous comment](https://github.com/OpenModelica/OpenModelica/discussions/8484#discussioncomment-2062492) 😄 

Anyway, I myself have a slight preference for not mixing up textual equations and graphical connections in the same model, but it's a matter of style and taste, it's not forbidden by the specification, and it's no big deal anyway. So, anything goes, as long as you don't have connect statements involving non-connectors.
